src/sync/sync_permission.cpp

```diff
@@ -0,0 +1,218 @@
+                                  const ConfigMaker& make_config)
+{
+    auto realm = Permissions::permission_realm(user, make_config);
+    auto session = SyncManager::shared().get_session(realm->config().path, *realm->config().sync_config);
+    auto results_wrapper = std::make_shared<NotificationWrapper<Results>>();
+    auto table = ObjectStore::table_for_object_type(realm->read_group(), "Permission");
+    results_wrapper->object = Results(std::move(realm), *table);
+
+    auto async = [results_wrapper, callback=std::move(callback)](auto ex) mutable {
+        Results& res = results_wrapper->object;
+        if (ex) {
+            callback(nullptr, ex);
+            results_wrapper.reset();
+        } else if (res.size() > 0) {
+            // We monitor the raw results. The presence of a `__management` Realm indicates
+            // that the permissions have been downloaded (hence, we wait until size > 0).
+            TableRef table = ObjectStore::table_for_object_type(res.get_realm()->read_group(), "Permission");
+            size_t col_idx = table->get_descriptor()->get_column_index("path");
+            auto query = !(table->column<StringData>(col_idx).ends_with("/__permission")
+                           || table->column<StringData>(col_idx).ends_with("/__management"));
+            callback(std::make_unique<PermissionResults>(res.filter(std::move(query))), nullptr);
+            results_wrapper.reset();
+        }
+    };
+    results_wrapper->token = results_wrapper->object.async(std::move(async));
+}
+
+void Permissions::set_permission(std::shared_ptr<SyncUser> user,
+                                 Permission permission,
+                                 PermissionChangeCallback callback,
+                                 const ConfigMaker& make_config)
+{
+    const auto realm_url = user->server_url() + permission.path;
+    auto realm = Permissions::management_realm(std::move(user), make_config);
+    auto object_wrapper = std::make_shared<NotificationWrapper<Object>>();
+    CppContext context;
+
+    // Write the permission object.
+    realm->begin_transaction();
+    object_wrapper->object = Object::create<util::Any>(&context, realm, *realm->schema().find("PermissionChange"), AnyDict{
+        { "id",         util::uuid_string() },
+        { "createdAt",  Timestamp(0, 0) },
+        { "updatedAt",  Timestamp(0, 0) },
+        { "userId",     permission.condition.user_id },
+        { "realmUrl",   realm_url },
+        { "mayRead",    permission.access != Permission::AccessLevel::None },
+        { "mayWrite",   permission.access == Permission::AccessLevel::Write || permission.access == Permission::AccessLevel::Admin },
+        { "mayManage",  permission.access == Permission::AccessLevel::Admin },
+    }, false);
+    realm->commit_transaction();
+
+    auto block = [object_wrapper, callback=std::move(callback)](CollectionChangeSet, std::exception_ptr ex) mutable {
```

Comment:
This could do with a comment explaining what it's doing. It takes a close reading to figure out when the notification is unregistered.